server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -478,6 +464,32 @@ public SendFileResult phase1(final IndexCommit snapshot, final Supplier<Integer>
         }
     }
 
+    boolean hasSameSyncId(Version indexCreatedVersion, Store.MetadataSnapshot source, Store.MetadataSnapshot target) {
```

Comment:
If the commit on the replica does not have sequence number yet, its local checkpoint and max_seq_no are [-1](https://github.com/elastic/elasticsearch/blob/6.7/server/src/main/java/org/elasticsearch/index/seqno/SequenceNumbers.java#L57-L58); thus recovery still utilizes the syncId if we haven't processed any operation with sequence numbers.

> Would it be an option to instead include localCheckpoint of last safe commit in the prepareForTranslogOperations message and store/validate this on the target node? Only if isSequenceNumberBasedRecovery==false though, since then we know that we must have an identical last safe-commit on target too.

Yes but handling BWC between 6.7.x and 7.0.0 would not be easy.

Reply:
> If the commit on the replica does not have sequence number yet, its local checkpoint and max_seq_no are -1; thus recovery still utilizes the syncId if we haven't processed any operation with sequence numbers.

As far as I can see, this is only true if no indexing occurred on the primary such that it also has local checkpoint and max_seq_no = -1. In step 4, we assume some primary is relocated to newer version (bound to happen at some point during the upgrade) and in step 5, we add more data to the shard(s), such that any primary on upgraded nodes  have sequence numbers (but replicas on non-upgraded do not). I may have missed a detail, but I think we cannot then do sync-id based recovery when a node holding a replica is upgraded in this scenario?

Reply:
> I think we cannot then do sync-id based recovery when a node holding a replica is upgraded in this scenario.

Yes, this is intention of this PR.

Reply:
good observation @henningandersen. I'm not sure how common that upgrade scenario is (I still think we should test for it), but I also can't think of any other solution than the one you proposed. That one has crazy back- and forward-compatibility implications though as we already released 7.0, and I would rather like to avoid too much craziness in 6.x, given that we will probably have to maintain that version for quite a while. My current inclination is to rather live with this full file-sync when upgrading from 5.x and proceed with the solution proposed in this PR.

@dnhatn can you add a test for @henningandersen's upgrade scenario?